### PR TITLE
Add full Lollipop compatibillity

### PR DESCRIPTION
**Pull request by dk-zero-cool**
Just ignorer the commit from yesterday, forgot to merge up before continuing working on it.

From the testing that I have done on my own Lollipop device, this seams to fix things and get most of it working. When I say most of it, I of cause mean that this does not fix the Native issues with Xposed Framework on Lollipop, but that is not an issue with the module so it does not really mater here. 

I also added some consistency between Lollipop and Pre-Lollipop ROM's. This change has been tested by me on Android 4.1, 4.3 and 5.1 AOSP/CM ROM's using the appropriate Xposed versions. The technique used in XPrivacy.java to create the consistency have however been tested on many other devices, as I use this in my own module. 

Not sure if you want to add your own structure and such to this, in which case this should provide you with the idea of how to make it work. But in any case, the current code will run on your supported devices, including Lollipop, with the exception of the Native hooks, if used with unsupported Xposed versions.

### CODE REVIEW

File: src/biz/bokhorst/xprivacy/PrivacyService.java [hunk 1]
```diff
@@ -2293,24 +2291,7 @@ private boolean isAMLocked(int uid) {
 	}
 
 	private Context getContext() {
-		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/PrivacyService.java]: I had left this, because I don't like to use privacy variable for compatibility reasons, but since this is now been done for Lollipop, you are right this can go away.

File: src/biz/bokhorst/xprivacy/XActivityManager.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ public String getClassName() {
 							listHook.add(new XActivityManager(act, null, null));
 						else
 							listHook.add(new XActivityManager(act, PrivacyManager.cSystem, null));
-				} else
+				} else if (!server)
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XActivityManager.java]: I try to prevent negated condition for readability, but it really doesn't matter much here.

File: src/biz/bokhorst/xprivacy/XAppWidgetManager.java [hunk 3]
```diff
@@ -42,12 +42,15 @@ public String getClassName() {
 		getInstalledProviders, getInstalledProvidersForProfile, Srv_getInstalledProviders, Srv_getInstalledProvidersForProfile
 	};
 
-	public static List<XHook> getInstances() {
+	public static List<XHook> getInstances(boolean services) {
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XAppWidgetManager.java]: Maybe 'services' should read 'server' for consistence with the rest, but it is an unimportant detail.

File: src/biz/bokhorst/xprivacy/XParam.java [hunk 4]
```diff
@@ -29,6 +29,11 @@ public static XParam fromXposed(MethodHookParam param) {
 		xparam.args = param.args;
 		xparam.mResult = param.getResult();
 		xparam.mThrowable = param.getThrowable();
+		
+		if (xparam.args == null) {
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XParam.java]: Why is this needed now?

File: src/biz/bokhorst/xprivacy/XPrivacy.java [hunk 5]
```diff
@@ -29,64 +28,107 @@
 	private static String mSecret = null;
 	private static List<String> mListHookError = new ArrayList<String>();
 	private static List<CRestriction> mListDisabled = new ArrayList<CRestriction>();
-
-	// http://developer.android.com/reference/android/Manifest.permission.html
-
-	static {
-		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP && mListDisabled.size() == 0) {
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: If you were wondering: I disabled this to prevent tripping SELinux (which could result in a bootloop) when reading the 'disabled' file on Lollipop (initZygote may not have access to /data/system on all devices).

File: src/biz/bokhorst/xprivacy/XPrivacy.java [hunk 6]
```diff
@@ -29,64 +28,107 @@
...
+		 */
+		try {
+			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+				XposedBridge.hookAllMethods(Class.forName("android.app.ActivityThread"), "systemMain", new XC_MethodHook() {
+					@Override
+					protected void afterHookedMethod(MethodHookParam param) throws Throwable {
+						final ClassLoader loader = Thread.currentThread().getContextClassLoader();
+						
+						try {
+							XposedBridge.hookAllConstructors(Class.forName("com.android.server.am.ActivityManagerService", false, loader), new XC_MethodHook() {
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: In general I try to avoid using XposedBridge methods when that is not necessary, so that Xposed can easily be replaced by something else when there is a need, but since this class bridges XPrivacy and Xposed, it is not too bad here.

File: src/biz/bokhorst/xprivacy/XPrivacy.java [hunk 7]
```diff
@@ -29,64 +28,107 @@
...
+		}
 	}
-
+	
 	public void handleLoadPackage(final LoadPackageParam lpparam) throws Throwable {
 		// Check for LBE security master
-		if (Util.hasLBE())
+		int uid = Process.myUid();
+		
+		if (Util.hasLBE() || mHasErr || uid <= Process.SYSTEM_UID)
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: People want to restrict Android too (only available after enabling an option with a big warning), so I am curious why you have added the last part of the condition.

File: src/biz/bokhorst/xprivacy/XPrivacy.java [hunk 8]
```diff
@@ -29,64 +28,107 @@
...
-	// Common
-	private static void init(String path) {
-		Util.log(null, Log.WARN, "Init path=" + path);
-
+	private void bootstrapZygote() throws Throwable {
 		// Generate secret
 		mSecret = Long.toHexString(new Random().nextLong());
+		
+		/*
+		 * Zygote can access this file (Even with SELinux enabled) as it is running as the outer root.
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: Are you very sure about /data/system be accessible by initZygote?
Not every ROM is AOSP and some manufacturers have very tight SELinux policies.
As said before, this can result in bootloops.

File: src/biz/bokhorst/xprivacy/XPrivacy.java [hunk 9]
```diff
@@ -404,14 +457,21 @@ public static void hookAll(List<XHook> listHook, ClassLoader classLoader, String
 	}
 
 	private static void hook(final XHook hook, ClassLoader classLoader, String secret) {
+		boolean doLog = PrivacyManager.cShell.equals(hook.getRestrictionName());
+		
+		if(doLog)Log.d("TESTING", "Core: Adding hook " + hook.getRestrictionName() + "[" + hook.getSpecifier() + "]");
```
Comment (M66B) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: I guess this have to be removed before building a new release.

**Comment by M66B** [comment]
This is all very good work, thanks!

File: src/biz/bokhorst/xprivacy/PrivacyService.java (hunk 1 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/PrivacyService.java]: With the changes I made, the AM instance is now parsed to the service in all Android versions and all of them contains the "mContext" property, which is one of the few things that Google keeps consistent across the OS.

File: src/biz/bokhorst/xprivacy/XActivityManager.java (hunk 2 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XActivityManager.java]: Yes I was not sure how you wanted this, there are many ways to make it, but a simple argument seamed fine. And later a found a few of your own hooks with a similar argument, so I stuck with it.

File: src/biz/bokhorst/xprivacy/XAppWidgetManager.java (hunk 3 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XAppWidgetManager.java]: Yes, I did not see that you had a few hooks with a similar argument to begin with.

File: src/biz/bokhorst/xprivacy/XParam.java (hunk 4 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XParam.java]: I got some errors in logcat about trying to count null array on xparam.args. It might be a small error in the Lollipop Xposed version where the Xposed args array is null. It did not harm I think, other than printing errors once awhile.

File: src/biz/bokhorst/xprivacy/XPrivacy.java (hunk 5 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: Okay, it did not harm on my device so I left it enabled. But if it causes problems for some, then... Not really sure what it does anyways, I am guessing that it allows the app to pre-disable features from boot?

File: src/biz/bokhorst/xprivacy/XPrivacy.java (hunk 6 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: Yes that was my thought as well. I know that you are using custom classes in the hook files.

File: src/biz/bokhorst/xprivacy/XPrivacy.java (hunk 7 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: Maybe I read something wrong in the original file? But it looked to me as the only thing you used "android" for, was to add a hook in AM for Lollipop? If not, then my bad :)

File: src/biz/bokhorst/xprivacy/XPrivacy.java (hunk 8 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: I will not bet my life on it, as I can only test AOSP. I would just have thought so, since Zygote is running as outer root, but then again, you can never know with SELinux.

File: src/biz/bokhorst/xprivacy/XPrivacy.java (hunk 9 above)
Comment (dk-zero-cool) [inline src/biz/bokhorst/xprivacy/XPrivacy.java]: Ups, missed one :s

**Comment by dk-zero-cool** [comment]
Let's just hope this works for everyone. There are many different devices with many different Android code bases.

**Comment by hackel** [comment]
Big thanks to @dk-zero-cool! (Sorry for comment spasm, he/she deserves it!)

**Comment by banderos101** [comment]
@dk-zero-cool 

Yes, very very thankyou much dk-zero-cool for helping out
